### PR TITLE
[CssSelector] Fix escape patterns

### PR DESCRIPTION
**Pull request by fancyweb**
| Q             | A
| ------------- | ---
| Branch?       | 5.4
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Tickets       | https://github.com/symfony/symfony/issues/46682 and https://github.com/symfony/symfony/issues/44516
| License       | MIT
| Doc PR        | -

I rechecked the original Python source code we "borrowed" (https://github.com/scrapy/cssselect/blob/master/cssselect/parser.py) and fixed some issues.

### CODE REVIEW

File: src/Symfony/Component/CssSelector/Parser/Tokenizer/TokenizerPatterns.php [hunk 1]
```diff
@@ -49,22 +49,22 @@ public function __construct()
         $this->identifierPattern = '-?(?:'.$this->nmStartPattern.')(?:'.$this->nmCharPattern.')*';
         $this->hashPattern = '#((?:'.$this->nmCharPattern.')+)';
         $this->numberPattern = '[+-]?(?:[0-9]*\.[0-9]+|[0-9]+)';
-        $this->quotedStringPattern = '([^\n\r\f%s]|'.$this->stringEscapePattern.')*';
+        $this->quotedStringPattern = '([^\n\r\f\\\\%s]|'.$this->stringEscapePattern.')*';
```
Comment (fancyweb) [inline src/Symfony/Component/CssSelector/Parser/Tokenizer/TokenizerPatterns.php:52]: https://github.com/scrapy/cssselect/blob/ddd9784977fca6e7da4439d837e4f510f1f10638/cssselect/parser.py#L926 (We are missing `\\` before the quote)

**Review by stof — APPROVED** [review]

File: src/Symfony/Component/CssSelector/Parser/Tokenizer/TokenizerPatterns.php [hunk 2]
```diff
@@ -49,22 +49,22 @@ public function __construct()
...
     }
 
     public function getNewLineEscapePattern(): string
     {
-        return '~^'.$this->newLineEscapePattern.'~';
+        return '~'.$this->newLineEscapePattern.'~';
```
Comment (fancyweb) [inline src/Symfony/Component/CssSelector/Parser/Tokenizer/TokenizerPatterns.php:57]: https://github.com/scrapy/cssselect/blob/ddd9784977fca6e7da4439d837e4f510f1f10638/cssselect/parser.py#L930

Check the 3 escape`"_sub_*` "functions" and how they are used. All occurrences are replaced.

File: src/Symfony/Component/CssSelector/Tests/Parser/ParserTest.php [hunk 3]
```diff
@@ -138,6 +138,16 @@ public function getParserTestData()
             ['div:not(div.foo)', ['Negation[Element[div]:not(Class[Element[div].foo])]']],
             ['td ~ th', ['CombinedSelector[Element[td] ~ Element[th]]']],
             ['.foo[data-bar][data-baz=0]', ["Attribute[Attribute[Class[Element[*].foo][data-bar]][data-baz = '0']]"]],
+            ['div#foo\.bar', ['Hash[Element[div]#foo.bar]']],
+            ['div.w-1\/3', ['Class[Element[div].w-1/3]']],
+            ['#test\:colon', ['Hash[Element[*]#test:colon]']],
+            [".a\xc1b", ["Class[Element[*].a\xc1b]"]],
+            // unicode escape: \22 == "
+            ['*[aval="\'\22\'"]', ['Attribute[Element[*][aval = \'\'"\'\']]']],
```
Comment (fancyweb) [inline src/Symfony/Component/CssSelector/Tests/Parser/ParserTest.php:146]: I borrowed those tests from https://github.com/scrapy/cssselect/blob/ddd9784977fca6e7da4439d837e4f510f1f10638/tests/test_cssselect.py#L550. They highlighted the "quotedStringPattern" problem.

**Comment by stof** [comment]
@fancyweb if you have some availability and are keen on checking tat python code, you might want to work on porting the missing features to Symfony 6.3 :smile: I create a few issues about them (but I'm quite sure the list is still incomplete for now)

**Review by fabpot — APPROVED** [review]
Having a followup PR that would catch up with upstream would be great indeed!

**Comment by fabpot** [comment]
Thank you @fancyweb.

**Pull request merged**